flocker/node/agents/blockdevice.py

```diff
@@ -1503,274 +1751,98 @@ def _mountpath_for_dataset_id(self, dataset_id):
         """
         return self.mountroot.child(dataset_id.encode("ascii"))
 
-    def calculate_changes(self, configuration, cluster_state, local_state):
-        this_node_config = configuration.get_node(
-            self.node_uuid, hostname=self.hostname)
-        local_node_state = cluster_state.get_node(self.node_uuid,
-                                                  hostname=self.hostname)
-
-        # We need to know applications (for now) to see if we should delay
-        # deletion or handoffs. Eventually this will rely on leases instead.
-        # https://clusterhq.atlassian.net/browse/FLOC-1425.
-        if local_node_state.applications is None:
-            return in_parallel(changes=[])
-
-        not_in_use = NotInUseDatasets(local_node_state, configuration.leases)
-
-        configured_manifestations = this_node_config.manifestations
-
-        configured_dataset_ids = set(
-            manifestation.dataset.dataset_id
-            for manifestation in configured_manifestations.values()
-            # Don't create deleted datasets
-            if not manifestation.dataset.deleted
+    def _calculate_desired_state(
+        self, configuration, local_applications, local_datasets
+    ):
+        not_in_use = NotInUseDatasets(
+            node_uuid=self.node_uuid,
+            applications=local_applications,
+            leases=configuration.leases,
         )
 
-        local_dataset_ids = set(local_node_state.manifestations.keys())
+        this_node_config = configuration.get_node(
+            self.node_uuid, hostname=self.hostname)
 
-        manifestations_to_create = set()
-        all_dataset_ids = list(
-            dataset.dataset_id
-            for dataset, node
-            in cluster_state.all_datasets()
-        )
-        for dataset_id in configured_dataset_ids.difference(local_dataset_ids):
-            if dataset_id in all_dataset_ids:
-                continue
+        desired_datasets = {}
+        for manifestation in this_node_config.manifestations.values():
+            dataset_id = UUID(manifestation.dataset.dataset_id)
+            # XXX: Make this configurable. FLOC-2679
+            maximum_size = manifestation.dataset.maximum_size
+            if maximum_size is None:
+                maximum_size = DEFAULT_DATASET_SIZE
+
+            if (
+                manifestation.dataset.deleted is True
+                # XXX FLOC-1772
+                and dataset_id in local_datasets
```

Comment:
I don't think you need to check local state or location; I think just checking for `deleted == True` is sufficient. See suggestions I made for changing the state transition table above.